### PR TITLE
feat: #997 — Auto-enable web search for capable models in Nexus

### LINKED ISSUE #997: [FS#160728] Weather Search on ChatGPT
Bug report from FreshService Ticket #160728

## Summary
Weather Search on ChatGPT — User asked for real-time weather in Nexus chat but the model responded it had no access to location data, expecting ChatGPT-style automatic web search.

## Description
User asked "What is the weather in Gig Harbor WA today?" in AI Studio Nexus. The AI responded: "I don't have access to the exact location behind that reference. Could you please tell me the city and state (or ZIP code) you'd like the weather for?" — despite the user having provided a specific named city. The ticket title "Weather Search on ChatGPT" indicates the user expected automatic web search capability comparable to ChatGPT.

Browser & system info from the report:
- Page URL: https://aistudio.psd401.ai/nexus?id=3cc139a8-9c89-4d70-a0b1-a416cc748ca3
- Page Title: AI Studio
- Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) Chrome/148.0.0.0
- Timestamp: 2026-05-21T19:53:59.683Z

## Ticket Information
- FreshService Ticket: #160728
- Status: Open
- Priority: Medium
- Urgency: 1
- Category: (none set)
- Created: 2026-05-21T19:54:00Z

## Reporter
- Name: Victoria Fairweather
- Contact: see FreshService #160728

## Routing reasoning
The page URL is `aistudio.psd401.ai/nexus`, the page title is "AI Studio", and the issue involves user-facing model interaction / tool availability in the Nexus chat feature. This is squarely in the AI Studio web app. Target: `psd401/aistudio`.

## Conversation History
No follow-up conversation on record — ticket was submitted and remains open with no replies.

---

### Suspected Root Cause — MEDIUM confidence

Web search is an **opt-in tool in Nexus that must be manually enabled** by the user via the Tools popover (wrench icon) before each conversation. It is not on by default. The user asked a real-time weather question without enabling web search, so the model fell back on its training data and produced the "I don't have access to the exact location" hedging response. This is working-as-designed behavior, but represents a significant UX/expectation mismatch versus ChatGPT, which enables web search automatically for capable models.

There is also a secondary issue: some models (Bedrock Claude) **cannot use web search at all**, and for other models the tool requires `webSearch` OR `grounding` capability flags to be set in the database for the web search toggle to even appear in the popover.

### Likely Affected Files

| File | Lines | Why |
|------|-------|-----|
| `app/(protected)/nexus/page.tsx` | 391–418 | `enabledTools` initializes to `[]` on fresh page load — no tools pre-enabled |
| `app/(protected)/nexus/_components/chat/tools-popover.tsx` | 80–188 | Tool toggle UI — web search is opt-in only, no default-on behavior |
| `lib/tools/tool-registry.ts` | 56–61 | `webSearch` requires `webSearch` OR `grounding` capability flag; neither explained to user |
| `lib/tools/client-tool-registry.ts` | 32–62 | Client-side tool availability gated on model capabilities |
| `lib/tools/provider-native-tools.ts` | 150–164 | `createBedrockNativeTools` returns empty ToolSet — Bedrock Claude has no web search path at all |
| `app/api/nexus/chat/route.ts` | 137–139, 154–164 | System prompt does not mention web search; empty `enabledTools` → no search tool created |
| `lib/streaming/provider-adapters/openai-adapter.ts` | 57–113 | Web search tool only created when `enabledTools` contains `'webSearch'` |
| `lib/streaming/provider-adapters/gemini-adapter.ts` | 106–109 | `getSupportedTools()` returns `['webSearch']` for Gemini models — so Gemini CAN use it, if enabled |

### Recent Related Commits

| SHA | Summary | Author / Date |
|-----|---------|---------------|
| `6c438aa` | fix(mantle-proxy): graceful rescue when model goes silent on user message | Kris Hagel / 2026-05-19 |
| `724bbbf` | feat(pii): per-type confidence overrides + Nexus prompt defence-in-depth | Kris Hagel / 2026-05-13 |
| `3c1d2ef` | feat(nexus): route Gemini Deep Research model through Interactions API (#917) | Kris Hagel / 2026-04-28 |

No commits in the recent history specifically modified web search tool enablement defaults or the system prompt messaging around search capability.

### Reproduction Status: REPRODUCED (static analysis — no live server)

No dev server was running during analysis. However the code path is unambiguous:

1. User navigates to `/nexus` — `enabledTools` initializes to `[]` (`nexus/page.tsx:391–418`)
2. User does NOT open the Tools popover and enable "Web Search"
3. Chat request sent with `enabledTools: []`
4. `createTools([])` in `nexus/chat/route.ts:154–164` creates only the universal `show_chart` tool — no web search
5. Model has no search tool → falls back to training data → produces "I don't have real-time access" response
6. Specific phrasing "exact location behind that reference" is characteristic of OpenAI reasoning models (o3/o4/GPT-5) hedging on real-time queries

### Open Questions for the Implementer

1. **Which model was the user on?** If they were on a Bedrock Claude model, web search is architecturally impossible today — that's a different fix than a UX one. If on GPT-5/Gemini, it's pure discoverability.
2. **Should web search be auto-enabled by default** for models that support it, rather than requiring explicit opt-in?
3. **Should the system prompt proactively inform users** that they can enable web search when they ask real-time questions (weather, news, scores)?
4. **Is there an in-chat affordance** (e.g., a suggested action card) when the model's response reveals it lacks real-time data?
5. **Gemini vs. OpenAI behavior:** Gemini models' `getSupportedTools()` already returns `webSearch`, so they could default to enabled — is that intentional or an oversight?

### Research Gaps

All three diagnosis agents ran successfully. Live reproduction was blocked because no dev server was running in the triage environment — all findings are from static code analysis. Confidence remains MEDIUM because the exact model the user selected is unknown; the specific failure path (Bedrock / OpenAI / Gemini) differs slightly per model family.

---

*Imported from FreshService: https://psd401.freshservice.com/a/tickets/160728*
*Triaged by routine: psd-triage*

### PR DESCRIPTION
**Pull request by krishagel**
## Summary

- Fixes [FS#160728] where a user asked for real-time weather in Nexus chat and the model said it had no search access — because web search is opt-in and wasn't enabled
- Auto-enables `webSearch` in `ToolsPopover` when: (a) the model supports it (`webSearch` or `grounding` capability flag) and (b) no tools are currently enabled
- Respects explicit user choices — if the user disables web search, it stays off until the next model switch (which triggers a full page reload, resetting state)
- Bedrock Claude and other models without web search capability are unaffected (auto-enable only fires when the model advertises the capability)

## Changes

- `app/(protected)/nexus/_components/chat/tools-popover.tsx` — add `else if` branch in the model-load effect to auto-enable `webSearch` when `enabledTools` is empty and the model supports it
- `tests/components/tools-popover.test.tsx` — unit tests covering the auto-enable behavior (5 cases: auto-enable, no-support model, user-has-tools, invalid-tool-cleanup, no-model)

## Test Plan

- [x] Unit tests written for all branches of the auto-enable logic
- [x] Existing e2e spec `nexus-tools.spec.ts` unaffected — auto-enabling a checked box is idempotent
- [x] Security audit (PASSED — see audit comment)
- [ ] Human review

Closes #997

---
*Opened by the lfg routine.*

### CODE REVIEW

**Comment by krishagel** [comment]
## 🔒 Security Audit

**Scope**: `app/(protected)/nexus/_components/chat/tools-popover.tsx`

| Area | Finding | Severity |
|------|---------|----------|
| Injection (A03) | Auto-enabled tool name is hardcoded literal `'webSearch'`, no user input in the path | ✅ None |
| Access Control (A01) | Tool execution remains capability-gated server-side; UI state change cannot bypass server validation | ✅ None |
| SSRF | Web search was already available for capable models via opt-in; no new network surface exposed | ✅ None |
| Secret leakage | No credentials, tokens, or env vars touched | ✅ None |
| Auth/Session | No auth code changed; component lives inside `/(protected)` requiring active session | ✅ None |

**Result**: PASSED — no findings at any severity level.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/(protected)/nexus/_components/chat/tools-popover.tsx [hunk 1]
```diff
@@ -109,15 +109,19 @@ export function ToolsPopover({
...
+
         if (validEnabledTools.length !== currentEnabledTools.length) {
-            onToolsChangeRef.current(validEnabledTools)
-          }
+          // Some previously-enabled tools are no longer available for this model
+          onToolsChangeRef.current(validEnabledTools)
+        } else if (currentEnabledTools.length === 0 && availableToolNames.includes('webSearch')) {
+          // No tools enabled yet — auto-enable web search for models that support it
+          onToolsChangeRef.current(['webSearch'])
+        }
```
[the remark this answers is not in the document]
Comment (krishagel) [reply app/(protected)/nexus/_components/chat/tools-popover.tsx:138]: Fixed in ea90132. The `if/else if` structure has been reworked so that when stripping invalid tools leaves `validEnabledTools` empty _and_ webSearch is available, webSearch is auto-enabled rather than calling `onToolsChange([])`. A `supportsWebSearch` flag now gates both the strip-then-empty path and the fresh-empty path, keeping the two cases consistent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/components/tools-popover.test.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    )
+
+    await waitFor(() => {
+      // Should strip the invalid tool but not auto-add webSearch
+      expect(onToolsChange).toHaveBeenCalledWith([])
+    })
+
+    // Must NOT have been called with webSearch
+    expect(onToolsChange).not.toHaveBeenCalledWith(['webSearch'])
+  })
```
[the remark this answers is not in the document]
Comment (krishagel) [reply tests/components/tools-popover.test.tsx]: Fixed in ea90132. Test case 4 is now `'removes invalid tools when switching models and auto-enables webSearch when result is empty'` and asserts `onToolsChange` is called with `['webSearch']` rather than `[]`. Two additional test cases were also added: one covering `disableAutoEnable` (no auto-enable in prompt library contexts) and one verifying no double-fire when webSearch is already enabled.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/(protected)/nexus/_components/chat/tools-popover.tsx [hunk 3]
```diff
@@ -109,15 +109,19 @@ export function ToolsPopover({
...
         )
+
         if (validEnabledTools.length !== currentEnabledTools.length) {
-            onToolsChangeRef.current(validEnabledTools)
-          }
+          // Some previously-enabled tools are no longer available for this model
+          onToolsChangeRef.current(validEnabledTools)
+        } else if (currentEnabledTools.length === 0 && availableToolNames.includes('webSearch')) {
+          // No tools enabled yet — auto-enable web search for models that support it
+          onToolsChangeRef.current(['webSearch'])
```
[the remark this answers is not in the document]
Comment (krishagel) [reply app/(protected)/nexus/_components/chat/tools-popover.tsx]: Fixed in ea90132. Added a `disableAutoEnable?: boolean` prop (defaults to `false`). Both `prompt-library/new/prompt-create-form.tsx` and `prompt-library/[id]/prompt-edit-form.tsx` now pass `disableAutoEnable` to `ToolsPopover`, so selecting a webSearch-capable model in the prompt editor won't silently persist `tools: ['webSearch']` on prompts the user never opted into tools for. The Nexus composer is unaffected (prop omitted = auto-enable stays on).

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (krishagel) [reply app/(protected)/nexus/_components/chat/tools-popover.tsx]: Fixed in ea90132. Added a `cancelled` boolean ref with a cleanup function (`return () => { cancelled = true }`). If the model changes before the in-flight `getAvailableToolsForModel` promise resolves, the `.then()` callback checks `if (cancelled) return` and exits without calling `onToolsChange` or updating state. The `.finally()` is similarly guarded so `setIsLoading(false)` is skipped for stale responses.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 1 complete.

- Review threads addressed: 4 (all resolved)
- CI failures fixed: 0 (CI was green)
- Commits pushed: 1 (`ea90132`)

**Changes in this round:**

| Issue | Fix |
|-------|-----|
| Logic bug: strip-all-tools + webSearch available → nothing enabled | Now auto-enables webSearch when strip leaves empty set and model supports it |
| CLAUDE.md violation: 9-line multi-line comment block in test file | Collapsed to single-line comment |
| Scope: auto-enable fires in prompt library forms | Added `disableAutoEnable` prop; both prompt-library forms now pass it |
| Stale response race: model switch before promise resolves | Added `cancelled` flag with cleanup function |
| Updated test case 4 | Now asserts `['webSearch']` after strip (matches corrected logic) |
| Added 2 new test cases | `disableAutoEnable` guard + webSearch-already-enabled no-op |

**Deferred (flagged for human review):**
- User preference resets on every model round-trip (suggested: `userDisabledWebSearch` ref) — behavioral product decision
- Hardcoded `'webSearch'` string → `WEB_SEARCH_TOOL_NAME` constant was added to the component; the test fixture still has its own local `'webSearch'` string literal which is fine for test isolation



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 2 complete.

- Review comments addressed: 1 (missing test case)
- CI failures fixed: 0 (CI was green)
- Commits pushed: 1 (`36483db`)

**Changes in this round:**

| Issue | Fix |
|-------|-----|
| Missing test: strip-to-empty without webSearch available | Added test case `'removes invalid tools when switching models and calls onToolsChange([]) when webSearch is not available'` — covers the `finalTools = validEnabledTools = []` code path (`supportsWebSearch = false`). Previously this branch was untested. |

**Deferred (flagged for human decision):**
- User preference resets on every model round-trip — behavioral/product decision whether to add a `userDisabledWebSearch` ref; not a correctness bug
- `disableAutoEnable` in effect deps list fires an extra fetch if the prop ever becomes dynamic — low-risk, noted for awareness



---
_Generated by [Claude Code](https://claude.ai/code)_
